### DISCUSSION TITLE
question: about Gaussblur min_ampl

### DISCUSSION THREAD

**Discussion by jishi92** [libvips/libvips#3038] [Q&A]
what's mean of the `min_ampl` ,I know Gaussblur has `radius` of convolution kernel and sigma of the Gaussian function,
But,What am I supposed to make of this min_ampl.
If I want to rewrite Gaussblur form imagemagick to libvips, how to  align the parameters.

**Comment by jcupitt** [comment]
Hi @qisenshi,

It's the minimum amplitude. Gaussian filters are infinite, so when you make a discrete mask, you need to decide when to stop. Setting `min_ampl` to 0.1 (for example) means you stop when the mask drops below 0.1. It's like radius, but it's easier to control because it doesn't depend on sigma.

For example:

```
$ python3
Python 3.10.4 (main, Jun 29 2022, 12:14:53) [GCC 11.2.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import pyvips
>>> pyvips.Image.gaussmat(3, 0.1, precision="float", separable=True).rot90().numpy()
array([[0.13533528],
       [0.24935221],
       [0.41111229],
       [0.60653066],
       [0.8007374 ],
       [0.94595947],
       [1.        ],
       [0.94595947],
       [0.8007374 ],
       [0.60653066],
       [0.41111229],
       [0.24935221],
       [0.13533528]])
>>> 
```

So that's making a radius 7 mask. If I try 0.2:

```
>>> pyvips.Image.gaussmat(3, 0.2, precision="float", separable=True).rot90().numpy()
array([[0.24935221],
       [0.41111229],
       [0.60653066],
       [0.8007374 ],
       [0.94595947],
       [1.        ],
       [0.94595947],
       [0.8007374 ],
       [0.60653066],
       [0.41111229],
       [0.24935221]])
>>> 
```

Now it's radius 5. It'll be slightly less accurate, but will blur by the same amount. 0.01 will be bigger:

```
>>> pyvips.Image.gaussmat(3, 0.01, precision="float", separable=True).rot90().numpy()
array([[0.011109  ],
       [0.0285655 ],
       [0.06572853],
       [0.13533528],
       [0.24935221],
       [0.41111229],
       [0.60653066],
       [0.8007374 ],
       [0.94595947],
       [1.        ],
       [0.94595947],
       [0.8007374 ],
       [0.60653066],
       [0.41111229],
       [0.24935221],
       [0.13533528],
       [0.06572853],
       [0.0285655 ],
       [0.011109  ]])
>>> 
```

Now it's radius 9. It'll be very accurate, but also rather slow.

**Reply by jishi92** [reply]
Hi jcupitt , thanks for your answer and examples. I think I got it. 
Is there any equation can calculate min_ampl by radius ?

**Reply by jcupitt** [reply]
I wouldn't bother, just use a fixed value for `min_ampl`, such as 0.1.

It's much better than radius -- you don't need to change it with sigma, you can just use the same value everywhere, depending on the accuracy/speed tradeoff you want.

**Reply by jishi92** [reply]
I konw,but my program used the Magic library before and it has the radius and  sigma.
Now I want to migrate it to libvips. 
I don't know what to do with these two parameters, to ensure users are not affected.

**Reply by jcupitt** [reply]
I understand, but you don't need to translate radius, just use a fixed value of `min_ampl`.

The radius parameter in magick sets the speed of the operation by setting the mask size, and this indirectly affects the accuracy, depending on the value of sigma.

With libvips, you can set the desired accuracy and it automatically computes the correct radius for you. 0.1 is very accurate, 0.2 is probably good enough.

**Reply by jishi92** [reply]
OK, I will test 0.1 and ` exp( - ((double)(x * x)) / sig2 )`, then chose a more accurate value.
Thanks for everything, best wishes！
